### PR TITLE
Feature/text improvements

### PR DESCRIPTION
**Pull request by TobiasRoikjer**
Merge with https://github.com/calabash/calabash-android-server/pull/10

### CODE REVIEW

File: ruby-gem/lib/calabash-android/text_helpers.rb [hunk 1]
```diff
@@ -34,22 +36,26 @@ def clear_text_in(query_string, options={})
...
-          puts "Warning: The method clear_text now clears the text in the currently focused view. Use clear_text_in instead"
-          puts "Notice that clear_text_in only clears the text of the first element matching the given query, not all."
-          puts "Use query(query, setText: '') to replicate the old behaviour"
-
-          clear_text_in(options)
-        else
-          perform_action('clear_text')
-        end
+        set_selection(-1, -1)
+        perform_action("delete_surrounding_text", -1, 0)
```
Comment (sapieneptus) [inline ruby-gem/lib/calabash-android/text_helpers.rb]: @TobiasRoikjer I'm not sure I see the advantage of this over `editable.clear()`.

Comment (sapieneptus) [inline ruby-gem/lib/calabash-android/text_helpers.rb]: As I implied in the corresponding calabash-android-server repo PR, I'm not a huge fan of having -1 as a sentinel value to indicate that it should be the end of the string (i.e., why not -2? ). I'd probably vote for a string constant value, like `"TEXT_END"` or something. Since they just get passed as strings anyway, it'd be easy to check for that case in the server and would be more readable.

File: ruby-gem/lib/calabash-android/text_helpers.rb [hunk 2]
```diff
@@ -20,8 +20,10 @@ def keyboard_enter_char(character, options = {})
         keyboard_enter_text(character[0,1], options)
       end
 
+      # Appends `text` into the first view matching `uiquery`.
       def enter_text(uiquery, text, options = {})
         tap_when_element_exists(uiquery, options)
+        set_selection(-1, -1)
```
Comment (sapieneptus) [inline ruby-gem/lib/calabash-android/text_helpers.rb]: See comments below regarding -1 as a sentinel value.

File: ruby-gem/lib/calabash-android/text_helpers.rb [hunk 3]
```diff
@@ -34,22 +36,26 @@ def clear_text_in(query_string, options={})
...
+      #
+      # @param [Integer] selection_start The start of the selection, can be
+      #  negative to begin counting from the end of the string.
+      # @param [Integer] selection_end The end of the selection, can be
+      #  negative to begin counting from the end of the string.
+      def set_selection(selection_start, selection_end)
+        perform_action("set_selection", selection_start, selection_end)
+      end
+
       def keyboard_visible?
```
Comment (sapieneptus) [inline ruby-gem/lib/calabash-android/text_helpers.rb]: I realize this is old and not part of this PR, but is there a way to do this in the android server instead of clientside?

File: ruby-gem/lib/calabash-android/text_helpers.rb (hunk 1 above)
Comment (TobiasRoikjer) [inline ruby-gem/lib/calabash-android/text_helpers.rb]: 1) The old clear_text action was not implemented as `editable.clear()`.
2) `editable.clear()` is not the right function to call :) This is how the default android keyboard is implemented

-1 does not just mean text end, it means the last element, -2 means the second to last element. This is the same as various other programming language conventions.

File: ruby-gem/lib/calabash-android/text_helpers.rb (hunk 3 above)
Comment (TobiasRoikjer) [inline ruby-gem/lib/calabash-android/text_helpers.rb]: I don't quite understand what you are trying to say here

Comment (sapieneptus) [inline ruby-gem/lib/calabash-android/text_helpers.rb]: I recently implemented a `KeyboardVisible()` function in UITest which uses the same `adb` command. I'm wondering if there isn't a way to implement that check server-side and have both clients just make a call to the server to find out if the keyboard is shown.

File: ruby-gem/lib/calabash-android/text_helpers.rb (hunk 1 above)
Comment (sapieneptus) [inline ruby-gem/lib/calabash-android/text_helpers.rb]: I'm not sure I agree about `editable.clear()`. Calling `inputConnection.deleteSurroundingText()` seems like more of a hack compared to `editable.clear()`, which is defined in the android docs to do exactly what we are trying to do. 

Put another way, asking the `editable` to clear itself seems more reliable than asking the `inputConnection` to clear the editable. 

Also, can you explain the params to `perform_action("delete_surrounding_text", -1, 0)`  being -1, 0 as opposed to 0, 0 ? Even if you implemented `ClearText.java` as just a convenience wrapper to `DeleteSurroundingText`, wouldn't it be more understandable to have `clear_text` calling `perform_action("clear_text")`?

File: ruby-gem/lib/calabash-android/text_helpers.rb (hunk 3 above)
Comment (TobiasRoikjer) [inline ruby-gem/lib/calabash-android/text_helpers.rb]: Which check?

Comment (TobiasRoikjer) [inline ruby-gem/lib/calabash-android/text_helpers.rb]: This PR does not introduce keyboard visibility

File: ruby-gem/lib/calabash-android/text_helpers.rb (hunk 1 above)
Comment (TobiasRoikjer) [inline ruby-gem/lib/calabash-android/text_helpers.rb]: I don't know what the Android docs say, the keyboard functions by calling these exact methods, with the arguments used in the client

File: ruby-gem/lib/calabash-android/text_helpers.rb (hunk 3 above)
Comment (sapieneptus) [inline ruby-gem/lib/calabash-android/text_helpers.rb]: I know, hence "I realize this is old and not part of this PR, but ..."

I'm just wondering if in theory that could be done easily in the server.

File: ruby-gem/lib/calabash-android/text_helpers.rb (hunk 1 above)
Comment (sapieneptus) [inline ruby-gem/lib/calabash-android/text_helpers.rb]: Then perhaps you could demonstrate via an XTC run that `delete_surrounding_text` will consistently clear textfields, even with autocorrect enabled?

File: ruby-gem/lib/calabash-android/text_helpers.rb (hunk 3 above)
Comment (TobiasRoikjer) [inline ruby-gem/lib/calabash-android/text_helpers.rb]: It can't :) The reason why your dumpsys implementation is perfect is because the service that actually determines when to show the keyboard reports the data
